modules/negotiate.go

```diff
@@ -0,0 +1,260 @@
+		// host would add 2 siacoins of funding to the contract.
+		//
+		// MaxCollateral indicates the maximum number of coins that a host is
+		// willing to put into a file contract.
+		Collateral         types.Currency `json:"collateral"`
+		CollateralFraction types.Currency `json:"collateralfraction"`
+		MaxCollateral      types.Currency `json:"maxcollateral"`
+
+		ContractPrice          types.Currency `json:"contractprice"`
+		DownloadBandwidthPrice types.Currency `json:"downloadbandwidthprice"`
+		StoragePrice           types.Currency `json:"storageprice"`
+		UploadBandwidthPrice   types.Currency `json:"uploadbandwidthprice"`
+
+		// Because the host has a public key, and settings are signed, and
+		// because settings may be MITM'd, settings need a revision number so
+		// that a renter can compare multiple sets of settings and determine
+		// which is the most recent.
+		RevisionNumber uint64 `json:"revisionnumber"`
+		Version        string `json:"version"`
+	}
+
+	// A RevisionAction is a description of an edit to be performed on a
+	// contract.
+	RevisionAction struct {
+		Type        types.Specifier
+		SectorIndex uint64
+		Offset      uint64
+		Data        []byte
+	}
+)
+
+// CreateAnnouncement will take a host announcement and encode it, returning
+// the exact []byte that should be added to the arbitrary data of a
+// transaction.
+func CreateAnnouncement(addr NetAddress, pk types.SiaPublicKey, sk crypto.SecretKey) (signedAnnouncement []byte, err error) {
+	// Create the HostAnnouncement and marshal it.
+	annBytes := encoding.Marshal(HostAnnouncement{
+		Specifier:  PrefixHostAnnouncement,
+		NetAddress: addr,
+		PublicKey:  pk,
+	})
+
+	// Create a signature for the announcement.
+	annHash := crypto.HashBytes(annBytes)
+	sig, err := crypto.SignHash(annHash, sk)
+	if err != nil {
+		return nil, err
+	}
+	// Return the signed announcement.
+	return append(annBytes, sig[:]...), nil
+}
+
+// DecodeAnnouncement decodes announcement bytes into a host announcement,
+// verifying the prefix and the signature.
+func DecodeAnnouncement(fullAnnouncement []byte) (ha HostAnnouncement, err error) {
+	// Read the first part of the announcement to get the intended host
+	// announcement.
+	r := bytes.NewReader(fullAnnouncement)
+	err = encoding.NewDecoder(r).Decode(&ha)
+	if err != nil {
+		return HostAnnouncement{}, err
+	}
+
+	// Check that the announcement was registered as a host announcement.
+	if ha.Specifier != PrefixHostAnnouncement {
+		return HostAnnouncement{}, ErrAnnNotAnnouncement
+	}
+	// Check that the public key is a recognized type of public key.
+	if ha.PublicKey.Algorithm != types.SignatureEd25519 {
+		return HostAnnouncement{}, ErrAnnUnrecognizedSignature
+	}
+
+	// Read the signature out of the reader.
+	var sig crypto.Signature
+	_, err = r.Read(sig[:])
+	if err != nil {
+		return HostAnnouncement{}, err
+	}
+	// Verify the signature.
+	var pk crypto.PublicKey
+	copy(pk[:], ha.PublicKey.Key)
```

Comment:
would be nice if we had a method for converting a `types.SiaPublicKey` to a `crypto.PublicKey`. It would check both the specifier and the length of the key slice.